File: tools/lib/card-design.mjs
import { existsSync, readFileSync } from "node:fs";
import { isAbsolute, relative, resolve, sep } from "node:path";

import yaml from "js-yaml";

export const CARD_DESIGN_MANIFEST = "templates/card-design/manifest.yaml";

const clone = value => structuredClone(value);

function inside(root, path) {
  const rel = relative(root, path);
  return rel === "" || (rel !== ".." && !rel.startsWith(`..${sep}`) && !isAbsolute(rel));
}

function readYaml(path) {
  return yaml.load(readFileSync(path, "utf8")) || {};
}

function designFile(gameDir, rel, label) {
  if (typeof rel !== "string" || !rel.startsWith("templates/card-design/") || isAbsolute(rel))
    throw new Error(`${label} must be under templates/card-design/: ${rel}`);
  const root = resolve(gameDir), path = resolve(root, rel);
  if (!inside(resolve(root, "templates/card-design"), path))
    throw new Error(`${label} escapes templates/card-design/: ${rel}`);
  if (!existsSync(path)) throw new Error(`${label} does not exist: ${rel}`);
  return path;
}

function dig(value, path) {
  for (const key of String(path || "").split(".").filter(Boolean)) {
    if (value == null || typeof value !== "object") return undefined;
    value = value[key];
  }
  return value;
}

export function cardMatchesFamily(card, match = {}) {
  return Object.entries(match).every(([path, wanted]) => {
    const choices = Array.isArray(wanted) ? wanted : [wanted];
    return choices.some(value => Object.is(dig(card, path), value));
  });
}

function fragment(gameDir, rel, label) {
  const doc = readYaml(designFile(gameDir, rel, label));
  if (!Array.isArray(doc.regions)) throw new Error(`${label} needs a regions array: ${rel}`);
  return doc;
}

function orderedRegions(regionMap, order) {
  const rank = new Map((order || []).map((id, index) => [id, index]));
  return [...regionMap.values()].sort((a, b) =>
    (rank.get(a.id) ?? Number.MAX_SAFE_INTEGER) - (rank.get(b.id) ?? Number.MAX_SAFE_INTEGER)
    || String(a.id).localeCompare(String(b.id)));
}

/** Load and compile a composable card-design catalog. The source fragments stay
 * first-class Git assets; compiled layouts are runtime-only values consumed by
 * the existing renderer. Later family regions override same-id components. */
export function loadCardDesign(gameDir, {
  normalizeLayout = value => value,
  manifestPath = CARD_DESIGN_MANIFEST,
} = {}) {
  if (!existsSync(resolve(gameDir, manifestPath))) return null;
  const resolvedManifestPath = designFile(gameDir, manifestPath, "card design manifest");
  const manifest = readYaml(resolvedManifestPath);
  if (manifest.version !== 1) throw new Error(`unsupported card design manifest version: ${manifest.version}`);
  if (!Array.isArray(manifest.families) || !manifest.families.length)
    throw new Error("card design manifest needs at least one family");
  const system = readYaml(designFile(gameDir, manifest.system, "design system"));
  if (!system.card || !Array.isArray(system.fonts) || !system.palette)
    throw new Error("design system needs card, fonts, and palette");
  const componentDocs = new Map();
  for (const component of manifest.components || []) {
    if (componentDocs.has(component.id)) throw new Error(`duplicate card design component: ${component.id}`);
    componentDocs.set(component.id, fragment(gameDir, component.source, `component ${component.id}`));
  }
  const familyIds = new Set();
  const families = manifest.families.map(family => {
    if (familyIds.has(family.id)) throw new Error(`duplicate card design family: ${family.id}`);
    familyIds.add(family.id);
    const regionMap = new Map();
    const componentIds = [];
    for (const component of manifest.components || []) {
      const applies = component.applies_to === "*" || component.applies_to?.includes(family.id);
      if (!applies) continue;
      componentIds.push(component.id);
      const doc = componentDocs.get(component.id);
      for (const id of doc.remove_regions || []) regionMap.delete(id);
      for (const region of doc.regions) regionMap.set(region.id, clone(region));
    }
    const familyDoc = fragment(gameDir, family.source, `family ${family.id}`);
    for (const id of familyDoc.remove_regions || []) regionMap.delete(id);
    for (const region of familyDoc.regions) regionMap.set(region.id, clone(region));
    const layout = normalizeLayout({ ...clone(system), regions: orderedRegions(regionMap, manifest.region_order) });
    return { ...clone(family), components: componentIds, region_count: layout.regions.length, layout };
  });
  return {
    version: manifest.version,
    name: manifest.name || "Card design families",
    description: manifest.description || "",
    manifest_path: manifestPath,
    system_path: manifest.system,
    legacy_source: manifest.legacy_source || null,
    components: (manifest.components || []).map(component => ({
      ...clone(component), region_count: componentDocs.get(component.id).regions.length,
    })),
    families,
    region_order: clone(manifest.region_order || []),
  };
}

export function resolveCardDesign(catalog, card) {
  if (!catalog || !card) return null;
  return catalog.families.find(family => cardMatchesFamily(card, family.match)) || null;
}

File: tools/card_design.py
"""Composable card-design loader shared by the hub builder and validators."""
from copy import deepcopy
from pathlib import Path

import yaml


MANIFEST = "templates/card-design/manifest.yaml"


def _inside(parent, child):
    try:
        child.resolve().relative_to(parent.resolve())
        return True
    except ValueError:
        return False


def _load(path):
    return yaml.safe_load(Path(path).read_text()) or {}


def _design_file(game_dir, rel, label):
    root = Path(game_dir).resolve()
    if not isinstance(rel, str) or not rel.startswith("templates/card-design/"):
        raise ValueError(f"{label} must be under templates/card-design/: {rel}")
    path = (root / rel).resolve()
    if not _inside(root / "templates/card-design", path):
        raise ValueError(f"{label} escapes templates/card-design/: {rel}")
    if not path.is_file():
        raise ValueError(f"{label} does not exist: {rel}")
    return path


def _fragment(game_dir, rel, label):
    doc = _load(_design_file(game_dir, rel, label))
    if not isinstance(doc.get("regions"), list):
        raise ValueError(f"{label} needs a regions array: {rel}")
    return doc


def source_get(value, path):
    for key in str(path or "").split("."):
        if not key:
            continue
        if not isinstance(value, dict):
            return None
        value = value.get(key)
    return value


def card_matches_family(card, match):
    for path, wanted in (match or {}).items():
        choices = wanted if isinstance(wanted, list) else [wanted]
        if source_get(card, path) not in choices:
            return False
    return True


def load_card_design(game_dir, normalize_layout=lambda value: value, manifest_path=MANIFEST):
    game_dir = Path(game_dir).resolve()
    if not (game_dir / manifest_path).exists():
        return None
    resolved_manifest_path = _design_file(game_dir, manifest_path, "card design manifest")
    manifest = _load(resolved_manifest_path)
    if manifest.get("version") != 1:
        raise ValueError(f"unsupported card design manifest version: {manifest.get('version')}")
    if not manifest.get("families"):
        raise ValueError("card design manifest needs at least one family")
    system = _load(_design_file(game_dir, manifest.get("system"), "design system"))
    if not system.get("card") or not isinstance(system.get("fonts"), list) or not system.get("palette"):
        raise ValueError("design system needs card, fonts, and palette")
    component_docs = {}
    for component in manifest.get("components") or []:
        if component["id"] in component_docs:
            raise ValueError(f"duplicate card design component: {component['id']}")
        component_docs[component["id"]] = _fragment(game_dir, component["source"], f"component {component['id']}")
    rank = {region_id: index for index, region_id in enumerate(manifest.get("region_order") or [])}
    families, seen = [], set()
    for family in manifest["families"]:
        if family["id"] in seen:
            raise ValueError(f"duplicate card design family: {family['id']}")
        seen.add(family["id"])
        region_map, component_ids = {}, []
        for component in manifest.get("components") or []:
            applies = component.get("applies_to")
            if applies != "*" and family["id"] not in (applies or []):
                continue
            component_ids.append(component["id"])
            doc = component_docs[component["id"]]
            for region_id in doc.get("remove_regions") or []:
                region_map.pop(region_id, None)
            for region in doc["regions"]:
                region_map[region["id"]] = deepcopy(region)
        family_doc = _fragment(game_dir, family["source"], f"family {family['id']}")
        for region_id in family_doc.get("remove_regions") or []:
            region_map.pop(region_id, None)
        for region in family_doc["regions"]:
            region_map[region["id"]] = deepcopy(region)
        regions = sorted(region_map.values(), key=lambda region: (rank.get(region["id"], 10**9), region["id"]))
        layout = normalize_layout({**deepcopy(system), "regions": regions})
        families.append({**deepcopy(family), "components": component_ids,
                         "region_count": len(regions), "layout": layout})
    return {
        "version": manifest["version"],
        "name": manifest.get("name") or "Card design families",
        "description": manifest.get("description") or "",
        "manifest_path": manifest_path,
        "system_path": manifest["system"],
        "legacy_source": manifest.get("legacy_source"),
        "components": [{**deepcopy(component), "region_count": len(component_docs[component["id"]]["regions"])}
                       for component in manifest.get("components") or []],
        "families": families,
        "region_order": deepcopy(manifest.get("region_order") or []),
    }


def resolve_card_design(catalog, card):
    if not catalog or not card:
        return None
    return next((family for family in catalog["families"]
                 if card_matches_family(card, family.get("match"))), None)
